wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/environment/InternalEnvironmentFactory.java

```diff
@@ -100,14 +103,17 @@ public T create(@Nullable final Environment sourceEnv)
           throw new InfrastructureException(e);
         }
 
-        machines.put(
-            machineEntry.getKey(),
+        InternalMachineConfig internalMachineConfig =
             new InternalMachineConfig(
                 installers,
                 normalizeServers(machineConfig.getServers()),
                 machineConfig.getEnv(),
                 machineConfig.getAttributes(),
-                machineConfig.getVolumes()));
+                machineConfig.getVolumes());
+
+        internalMachineConfig.getAttributes().put(CONTAINER_SOURCE_ATTRIBUTE, RECIPE);
```

Comment:
I think that in this case, you would not add an attribute to machines added by the environment factory implementation.

Reply:
@garagatyi Nice catch!
@AndrienkoAleksandr Please consider reworking it in the following way:
```java
    T internalEnv = doCreate(recipe, machines, warnings);

    internalEnv.getMachines()
        .values()
        .forEach(m -> m.getAttributes().put(CONTAINER_SOURCE_ATTRIBUTE, RECIPE));

    return internalEnv;
```
And consider adding a test case for it.

Reply:
Fixed